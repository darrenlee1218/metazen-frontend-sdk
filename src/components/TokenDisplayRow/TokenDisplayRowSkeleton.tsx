import React, { FC } from 'react';

import Skeleton from '@mui/material/Skeleton';
import Typography from '@mui/material/Typography';

import { TokenIcon } from '@components/TokenIcon';
import { boxCreator } from '@components/boxCreator';

const SkeletonRoot = boxCreator({
  p: 2,
  display: 'grid',
  gridTemplateColumns: '40px minmax(0, 1fr)',
  gap: 1.5,
  alignItems: 'center',
  borderRadius: 4,
  backgroundColor: (theme) => theme.palette.action.hover,
});

const TokenDetailBox = boxCreator({
  display: 'flex',
  flexDirection: 'column',
  justifyContent: 'center',
});

const TokenDetailRow = boxCreator({
  display: 'flex',
  alignItems: 'center',
  justifyContent: 'space-between',
  minWidth: 0,
  gap: 2,
});

export const TokenDisplayRowSkeleton: FC = () => {
  return (
    <SkeletonRoot>
      <Skeleton variant="circular">
        <TokenIcon isNative={false} tokenIconUrl="" networkIconUrl="" />
      </Skeleton>

      <TokenDetailBox>
        <TokenDetailRow>
          <Typography component="div" variant="h2" sx={{ flexGrow: 1 }}>
            <Skeleton />
          </Typography>

          <Typography component="div" variant="h2" sx={{ flexGrow: 1 }}>
            <Skeleton />
          </Typography>
        </TokenDetailRow>

        <TokenDetailRow>
          <Typography component="div" variant="h2" sx={{ flexGrow: 1 }}>
            <Skeleton />
          </Typography>

          <Typography component="div" variant="h2" sx={{ flexGrow: 1 }}>
            <Skeleton />
          </Typography>
        </TokenDetailRow>
      </TokenDetailBox>
    </SkeletonRoot>
  );
};
