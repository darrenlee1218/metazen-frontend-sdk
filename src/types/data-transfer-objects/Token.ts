interface TokenDisplay {
  /**
   * specific color for display in some places
   */
  colorCode: string;

  /**
   * URL of token icon
   */
  pngIconUrl: string;

  /**
   * URL of token icon
   */
  svgIconUrl: string;

  /**
   * NFT collection icon
   */
  nftCollectionImageUrl?: string;

  /**
   * background image for the NFT collection
   */
  nftCollectionBackgroundImageUrl?: string;
}

interface TokenChain {
  /**
   * unique chain ID
   */
  chainID: string;

  /**
   * name of the chain
   */
  chainName: string;

  /**
   * whether it's a testnet/mainnet chain
   */
  type: 'TESTNET' | 'MAINNET';

  /**
   * minimum block confirmatoin required for a transaction
   */
  minBlockConfirmation: number;

  /**
   * Smart contract URL
   */
  smartContractUrl: string;

  /**
   * Address URL (for reviewing wallet info)
   */
  addressUrl: string;

  /**
   * tx URL (for reviewing address history)
   */
  txUrl: string;
}

interface TokenNetwork {
  /**
   * network name for chainalysis usage
   */
  chainalysisNetwork: string;

  /**
   * color display for the network
   */
  networkColor: string;

  /**
   * URL of network icon
   */
  pngNetworkIconUrl: string;
}

// TODO: set Token['standard] = TokenStandard
export enum TokenStandard {
  NATIVE = 'NATIVE',
  ERC20 = 'ERC-20',
  ERC721 = 'ERC-721',
  ERC1155 = 'ERC-1155',
}

export default interface Token {
  /**
   * unique token ID, from chain ID, ticker and contract address (contract addr only for non-native token)
   */
  key: string;

  /**
   * token name
   */
  displayName: string;

  /**
   * asset ticker
   * WARNING: ticker may be not unique
   */
  ticker: string;

  /**
   * standard of an token (e.g. ERC-1155)
   * always empty for native token
   */
  standard: TokenStandard;

  /**
   * whether it is a native token
   */
  isNative: boolean;

  /**
   * token's Explorer address i.e.polygonscan
   */
  txUrl: string;

  /**
   * token's contract address
   * undefined for native token
   */
  contractAddress: string | undefined;

  /**
   * gas limit of token for estimate gas fee
   */
  gasLimit: string;

  /**
   * decimal place for token amount
   */
  decimal?: number;

  /**
   * network related stuff
   */
  network: TokenNetwork;

  /**
   * decimal places for custody
   * defined for all ERC-20 / Native tokens
   */
  custodyPrecision?: number;

  /**
   * decimal places for display
   * defined for all ERC-20 / Native tokens
   */
  displayPrecision?: number;
  /**
   * display-related properties of an token
   */
  display: TokenDisplay;

  /**
   * chain info
   */
  chain: TokenChain;
}
