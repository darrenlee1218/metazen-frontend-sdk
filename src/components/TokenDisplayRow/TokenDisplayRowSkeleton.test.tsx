import React from 'react';
import { render, fireEvent, screen, cleanup } from '@testing-library/react';
import { ComponentProps } from 'react';
import { TokenDisplayRowSkeleton } from './TokenDisplayRowSkeleton';
import { TestProviders } from '@utils/testing/TestProviders';
import Token from '@gryfyn-types/data-transfer-objects/Token';

afterEach(cleanup);

it('matches snapshot', () => {
  const { container } = render(<TokenDisplayRowSkeleton />, { wrapper: TestProviders });
  expect(container).toMatchSnapshot();
});
