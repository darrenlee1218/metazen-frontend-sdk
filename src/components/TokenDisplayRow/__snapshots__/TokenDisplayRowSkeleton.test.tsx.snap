// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`matches snapshot 1`] = `
<div>
  <div
    class="MuiBox-root css-mq31qy"
  >
    <span
      class="MuiSkeleton-root MuiSkeleton-circular MuiSkeleton-pulse MuiSkeleton-withChildren MuiSkeleton-fitContent MuiSkeleton-heightAuto css-gedcuj-MuiSkeleton-root"
    >
      <span
        class="MuiBadge-root BaseBadge-root css-1c32n2y-MuiBadge-root"
      >
        <div
          class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1b6koi5-MuiAvatar-root"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
            data-testid="PersonIcon"
            focusable="false"
            viewBox="0 0 24 24"
          >
            <path
              d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
            />
          </svg>
        </div>
        <span
          class="MuiBadge-badge MuiBadge-standard MuiBadge-anchorOriginTopRight MuiBadge-anchorOriginTopRightCircular MuiBadge-overlapCircular BaseBadge-badge css-6zrg6j-MuiBadge-badge"
        >
          <div
            class="MuiAvatar-root MuiAvatar-circular MuiAvatar-colorDefault css-1n5knhj-MuiAvatar-root"
            data-testid="network-icon"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiAvatar-fallback css-10mi8st-MuiSvgIcon-root-MuiAvatar-fallback"
              data-testid="PersonIcon"
              focusable="false"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 12c2.21 0 4-1.79 4-4s-1.79-4-4-4-4 1.79-4 4 1.79 4 4 4zm0 2c-2.67 0-8 1.34-8 4v2h16v-2c0-2.66-5.33-4-8-4z"
              />
            </svg>
          </div>
        </span>
      </span>
    </span>
    <div
      class="MuiBox-root css-zg1vud"
    >
      <div
        class="MuiBox-root css-qf05oz"
      >
        <div
          class="MuiTypography-root MuiTypography-h2 css-1vsyzh6-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1h2g3k5-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiTypography-root MuiTypography-h2 css-1vsyzh6-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1h2g3k5-MuiSkeleton-root"
          />
        </div>
      </div>
      <div
        class="MuiBox-root css-qf05oz"
      >
        <div
          class="MuiTypography-root MuiTypography-h2 css-1vsyzh6-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1h2g3k5-MuiSkeleton-root"
          />
        </div>
        <div
          class="MuiTypography-root MuiTypography-h2 css-1vsyzh6-MuiTypography-root"
        >
          <span
            class="MuiSkeleton-root MuiSkeleton-text MuiSkeleton-pulse css-1h2g3k5-MuiSkeleton-root"
          />
        </div>
      </div>
    </div>
  </div>
</div>
`;
